### PR TITLE
Feat: Add NodeJS E2EE TS example.

### PR DESCRIPTION
**Pull request by saul-jb**
This PR adds a NodeJS TypeScript example with E2EE enabled to the examples, this should help people figure out how to use the E2EE in a NodeJS environment.

E2EE is one of the primary reasons to use Matrix and the docs/examples are really scarce on how to handle it properly, this example is up to date and should really help with that.

Related issues:
* #2205
* #1726
* #437

---
This PR currently has none of the required changelog labels.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.

### CODE REVIEW

**Review by dbkr — COMMENTED** [review]
Hmm, at first glance this feels extremely complex for an example. In general, I'd expect a code example to be a single file of actual logic, probably limited to a few functions. The non-crypto node example is probably a bit too large, IMO. It's great to have a complete example but I'm struggling to see where the relevant code for starting an e2e Matrix client actually is in here.

I could also use some help understanding how this works. It's not specifying a store, so presumably is using the MemoryStore by default, which may be fine since it logs in each time it starts, but in an example it's even more important than normal to be very clear about what's happening.

The references to maybe clearing devices in the case of encryption problems don't really help me to understand when I actually would or wouldn't do this in real code.

File: examples/crypto-node/src/matrix.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+		.map(device => device.device_id)
+		.filter(id => id !== client.getDeviceId());
+
+	await Promise.all(devicesIds.map(id => client.deleteDevice(id)));
+};
+
+/**
+ * Start the client with a password login.
+ */
+export const start = async (passwordLogin: PasswordLogin, options?: { forgetDevices?: boolean }): Promise<MatrixClient> => {
```
[the remark this answers is not in the document]
Comment (saul-jb) [reply examples/crypto-node/src/matrix.ts]: Quite right, this was only used in testing it.

**Comment by saul-jb** [comment]
> Hmm, at first glance this feels extremely complex for an example.

I don't feel this example is too complicated, a little complex maybe but this allows me to segregate parts of the code that people want to look at without having to sift through the rest that they are not interested in like in `node` example.

Here we have `matrix-importer.ts` which handles the setup for importing the Matrix SDK correctly and `matrix.ts` which holds the methods for doing various actions through Matrix. The rest is mostly logic to turn it into a working example.

> I could also use some help understanding how this works.

I have updated the README to hopefully increase the clarity on how it is setup and what it is doing.

**Comment by saul-jb** [comment]
> All the other (few) examples concerning e2ee on node are outdated, difficult to find, and not working.

Yeah, Matrix is pretty cool overall but it is a nightmare to figure out how to use this library due to lack of working examples and documentation. I think a few up to date examples would go a long way to increasing adoption.

> Adding this section to the createClient of the startWithToken function doesn't work smoothly, it gives back and error such as

Yeah, that seems to have problems and the only way I have found to persist storage is using `node-localstorage`. I have updated the example to use this by default now.

> Also: if we want to have the crypto Store persisted, do we still need to login via username and password (getTokenLogin) ?

No, since we can persist an access token and device we can skip straight to the token login - the example has be updated to do this too.

**Comment by dbkr** [comment]
This came up as a longstanding PR in our meeting, and others on the team were in agreement with me that it's a bit complex for an example, although I also agree that we have a severe lack of examples and people are clearly finding this useful.

I don't think we'd want this to live in the main js-sdk repo still, but as a compromise, how about putting it in its own repo, if you'd be happy to maintain it? Linking to a separate repo as a usage example from the js-sdk docs seems like a thing we could do.

**Pull request closed without merging**